lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,354 @@
+}
+
+static const mlir::DialectRegistry &getDialectRegistry() {
+  static const mlir::DialectRegistry Registry = []() -> mlir::DialectRegistry {
+    mlir::DialectRegistry Registry;
+    // The DLTI dialect is used to express the data layout.
+    Registry.insert<mlir::DLTIDialect>();
+    // All dialects that implement the LLVMImportDialectInterface.
+    mlir::registerAllFromLLVMIRTranslations(Registry);
+    Registry.insert<mlir::clift::CliftDialect>();
+    return Registry;
+  }();
+  return Registry;
+}
+
+static ContextPtr makeContext() {
+  const auto Threading = MLIRContext::Threading::DISABLED;
+  return std::make_unique<MLIRContext>(getDialectRegistry(), Threading);
+}
+
+// Cloning MLIR from one module into another requires first serialising the
+// source and then deserialising it again into the target module. We do this
+// a) when cloning operations from one container to another, and
+// b) when "garbage collecting" types and attributes during remove.
+//
+// Ideally at least for use case a) we could clone from one context into another
+// directly, but that is not supported at this time.
+static OwningModuleRef cloneModuleInto(ModuleOp SourceModule,
+                                       MLIRContext &DestinationContext) {
+  llvm::SmallString<1024> Buffer;
+  llvm::raw_svector_ostream Out(Buffer);
+  mlir::writeBytecodeToFile(SourceModule, Out);
+
+  // Serialised MLIR contains multiple operations and as such must be
+  // deserialised into a block. We use a temporary block stored on stack.
+  mlir::Block OuterBlock;
+
+  const mlir::LogicalResult
+    R = mlir::readBytecodeFile(llvm::MemoryBufferRef(Buffer, "cloneModuleInto"),
+                               &OuterBlock,
+                               mlir::ParserConfig(&DestinationContext));
+  revng_assert(mlir::succeeded(R));
+
+  // The serialised code is known to contain exactly one ModuleOp operation.
+  auto &Operations = OuterBlock.getOperations();
+  revng_assert(Operations.size() == 1);
+  auto DestinationModule = mlir::cast<ModuleOp>(Operations.front());
+
+#ifndef NDEBUG
+  const size_t SrcOpsCount = getModuleOperations(SourceModule).size();
+  const size_t NewOpsCount = getModuleOperations(DestinationModule).size();
+  revng_assert(NewOpsCount == SrcOpsCount);
+#endif
+
+  // Detach the module from the temporary block before returning it.
+  DestinationModule->remove();
+
+  // The module is now owned by the OwningModuleRef returned to the caller.
+  return DestinationModule;
+}
+
+static pipeline::Target makeTarget(const MetaAddress &MA) {
+  return pipeline::Target(MA.toString(), kinds::MLIRFunctionKind);
+}
+
+static void makeExternal(FunctionOpInterface F) {
+  revng_assert(F->getNumRegions() == 1);
+
+  // A function is made external by clearing its region.
+  mlir::Region &Region = F->getRegion(0);
+  Region.dropAllReferences();
+  Region.getBlocks().clear();
+}
+
+// Remove any unused non-target symbols.
+static void pruneUnusedSymbols(ModuleOp Module) {
```

Comment:
Maybe [-symbol-dce](https://mlir.llvm.org/docs/Passes/#-symbol-dce) is relevant here?

Reply:
Hard to say off the cuff... It looks like the non-target functions would at least have to be marked private at some point.

Reply:
OK, maybe add a TODO comment about investigating this